### PR TITLE
fix(data-gen): reuse shared KV hashing for Mooncake IDs

### PR DESCRIPTION
**Pull request by RitwijParmar**
Part of #8997.

This removes one small source of hash-contract drift from `dynamo-data-gen`: `RollingHashIdMapper::hash_token_blocks` no longer reimplements its own local block + parent-chain hashing. It now derives sequence-aware hashes through `dynamo-kv-hashing::Request::sequence_hashes` and only keeps the existing data-gen responsibility of compacting stable sequence hashes into Mooncake `hash_ids`.

What changed:

- `dynamo-data-gen` depends on `dynamo-kv-hashing` instead of `bytemuck` + `dynamo-tokens`
- `hash_token_blocks` now uses the shared KV-hashing contract before compacting IDs
- added `try_hash_token_blocks` so invalid block sizes can be surfaced without panicking
- added regression coverage that compares token-block hashing against explicit `Request::sequence_hashes`
- added coverage for the universal contract's trailing partial-block behavior

One behavior note: trailing partial token blocks are no longer assigned Mooncake IDs by the token helper, matching `dynamo-kv-hashing`'s current complete-block-only contract. This is intentional here, but I called it out because it changes the old local helper's behavior.

Checks run locally:

- `cargo test -p dynamo-data-gen`
- `cargo test -p dynamo-kv-hashing`
- `cargo fmt --check --package dynamo-data-gen`
- `cargo clippy -p dynamo-kv-hashing -p dynamo-data-gen --all-targets -- -D warnings`
- `git diff --check`

I also tried the nearby consumer test:

- `cargo test -p dynamo-bench agent_trace --test agent_trace`

That did not reach the test phase locally because `etcd-client`'s build script requires `protoc`, which is not installed in this local environment.

AI disclosure: I used AI assistance while editing and reviewing this patch, then ran the checks above locally.

## Summary by CodeRabbit

* **New Features**
  * Introduced new fallible token block hashing API with explicit error handling and input validation. Existing APIs remain available and now delegate to the new error-handling implementation.

* **Tests**
  * Enhanced test suite with comprehensive coverage for token block hashing, including edge case validation, block-size constraint enforcement, and result consistency verification.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ai-dynamo/dynamo/pull/10095?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: lib/data-gen/src/mooncake.rs [hunk 1]
```diff
@@ -171,27 +175,30 @@ impl RollingHashIdMapper {
...
+}
+
+/// Fallible token-block hashing helper for callers that want to surface
+/// invalid block-size or request-shape errors.
+pub fn try_hash_token_blocks(mapper: &mut RollingHashIdMapper, tokens: &[u32]) -> Result<Vec<u64>> {
+    require_positive("block size", mapper.block_size)?;
+    let sequence_hashes = Request::builder()
+        .tokens(tokens.to_vec())
+        .build()?
+        .sequence_hashes(
```
Comment (dynamo-ops) [inline lib/data-gen/src/mooncake.rs]: `Request::sequence_hashes` drops trailing partial blocks, so `hash_token_blocks` now emits `floor(input_length / block_size)` IDs while Mooncake replay synthesis requires enough `hash_ids` to cover the full `input_length`, causing generated traces with non-multiple prompt lengths to be rejected or truncated by consumers. Fix: preserve the Mooncake capacity contract by appending a stable ID for the trailing partial block so `hash_ids.len() == tokens.len().div_ceil(block_size)`.

<details><summary>🤖 AI Fix</summary>

In `lib/data-gen/src/mooncake.rs`, update `try_hash_token_blocks` to keep using `dynamo_kv_hashing::Request::sequence_hashes` for complete blocks but append one deterministic mapped hash when `tokens.len() % mapper.block_size != 0`, and change `trailing_partial_block_is_not_mapped` to assert ceil-block output such as `[1,2,3] -> vec![0]` and `[1,2,3,4,5,6] -> vec![1,2]`.

</details>

Comment (hhzhang16) [reply lib/data-gen/src/mooncake.rs]: +1, I verified this against the data-gen -> mocker path: the exporter still writes full `input_length`, while replay capacity comes from `hash_ids.len() * trace_block_size`, so the floor-block behavior truncates or rejects non-multiple prompts.

**Comment by RitwijParmar** [comment]
I added follow-up coverage in 3e944cdd49 for the Mooncake wire format that this PR touches.

The new tests cover both upstream alias input (`input_tokens`, `output_tokens`, `created_time`, `delay_ms`) and canonical input, then verify serialization emits only canonical fields and that the derived `hash_ids` stay stable through the round trip. I kept this on `MooncakeRow` because `dynamo_kv_hashing::Request` itself is not a serde wire type in this branch.

Validated locally:
`cargo test -p dynamo-data-gen mooncake --lib` -> 21 passed

**Comment by RitwijParmar** [comment]
Small update: I amended the follow-up commit with DCO sign-off and force-pushed it as d56c339895. The code/test content is the same; this is just to clear the DCO check.

**Comment by RitwijParmar** [comment]
Fixed the partial-block capacity issue in `2d49173049`.\n\nThe Mooncake helper still uses `Request::sequence_hashes` for complete blocks, but now appends one stable mapped key for a trailing partial block. That keeps `hash_ids.len()` at ceil(input_length / block_size), so generated replay rows don’t lose capacity for non-multiple prompt lengths.\n\nI also updated the regression from the old floor-block expectation and added an exact-boundary test so full-block prompts don’t get an extra ID.\n\nValidated locally:\n- `cargo fmt --check --manifest-path lib/data-gen/Cargo.toml`\n- `cargo test -p dynamo-data-gen mooncake --lib` -> 22 passed

**Review by hhzhang16 — APPROVED** [review]

File: lib/data-gen/src/mooncake.rs [hunk 2]
```diff
@@ -171,27 +177,55 @@ impl RollingHashIdMapper {
...
+    complete_sequence_hashes: &[u64],
+) -> Option<u64> {
+    let tail_len = tokens.len() % mapper.block_size;
+    if tail_len == 0 {
+        return None;
+    }
+
+    let tail = &tokens[tokens.len() - tail_len..];
+    let parent = complete_sequence_hashes.last().copied();
+    let mut hasher = FxHasher::default();
```
Comment (PeaBrane) [inline lib/data-gen/src/mooncake.rs]: Could we avoid introducing a separate `FxHasher` contract for the tail? There is already a partial-block precedent in `lib/llm/src/agents/trace/replay.rs` (`input_sequence_hashes`, lines 48-65): it hashes the actual trailing token bytes as a truncated local block with `compute_hash_v2(..., XXH3_SEED)`, appends that local hash, and applies the existing rolling sequence-hash helper across the full list. Please reuse that contract here so Mooncake preserves replay capacity without making identical partial prompts hash differently depending on their producer.

**Review by PeaBrane — COMMENTED** [review]
One suggestion on keeping partial-block hashing aligned with the existing rolling hash contract.

File: lib/data-gen/src/mooncake.rs (hunk 2 above)
Comment (RitwijParmar) [reply lib/data-gen/src/mooncake.rs]: Thanks, that was the right contract to preserve. I pushed `75a355ab3f` to remove the separate `FxHasher` tail path.

The trailing partial block now:
- hashes the actual tail token bytes with `compute_hash_v2` using the same request salt
- chains it through `compute_next_sequence_hash(parent, tail_block_hash)` when a complete parent block exists
- keeps the root-partial case as the local tail block hash

I also added `trailing_partial_block_uses_shared_chain_contract` to compare the mapper output against a manually built complete-block plus chained-tail sequence.

Validated locally:
- `cargo fmt --check --manifest-path lib/data-gen/Cargo.toml`
- `cargo test -p dynamo-data-gen mooncake --lib` -> 23 passed

File: lib/data-gen/src/mooncake.rs [hunk 3]
```diff
@@ -171,27 +177,55 @@ impl RollingHashIdMapper {
...
-    }
-    hash_ids
+    try_hash_token_blocks(mapper, tokens).expect("Mooncake token-block hashing failed")
+}
+
+/// Fallible token-block hashing helper for callers that want to surface
+/// invalid block-size or request-shape errors.
+pub fn try_hash_token_blocks(mapper: &mut RollingHashIdMapper, tokens: &[u32]) -> Result<Vec<u64>> {
+    require_positive("block size", mapper.block_size)?;
+    let mut sequence_hashes = Request::builder()
```
Comment (PeaBrane) [inline lib/data-gen/src/mooncake.rs]: This now copies the full prompt into `Request`, and `Request::into_blocks` clones `self.tokens` again before hashing. Could we add a small consuming `Request::into_sequence_hashes(self, block_size)` API in `dynamo-kv-hashing` and use it here? That would preserve the existing borrowed API while moving the request tokens into block construction and eliminating the second O(prompt length) copy. A focused equivalence test between the borrowed and consuming paths should cover the new API.

**Review by PeaBrane — COMMENTED** [review]
One performance suggestion to avoid the additional full-prompt clone introduced by the shared hashing path.

File: lib/data-gen/src/mooncake.rs (hunk 3 above)
Comment (RitwijParmar) [reply lib/data-gen/src/mooncake.rs]: Good call. I pushed `b9b54fe4c1` to add a consuming `Request::into_sequence_hashes(self, block_size)` path in `dynamo-kv-hashing` and use it from Mooncake.

That keeps the borrowed `sequence_hashes(&self, ...)` API intact, but lets the one-shot Mooncake producer move the request tokens into `TokenBlockSequence` instead of cloning them again inside block construction.

I also added `consuming_sequence_hashes_match_borrowed_path` so the new API is covered against the existing borrowed projection.

Validated locally:
- `cargo fmt --check --manifest-path lib/data-gen/Cargo.toml`
- `cargo test -p dynamo-data-gen mooncake --lib` -> 23 passed
- `cargo test -p dynamo-kv-hashing consuming_sequence_hashes_match_borrowed_path --test request_hashing`
- `cargo clippy -p dynamo-data-gen --lib --tests -- -D warnings`
- `cargo clippy -p dynamo-kv-hashing --lib --tests -- -D warnings`

File: lib/data-gen/src/mooncake.rs (hunk 2 above)
Comment (PeaBrane) [reply lib/data-gen/src/mooncake.rs]: The functional shape looks right now. One dependency-boundary follow-up: could you re-export `compute_next_sequence_hash` from `dynamo-kv-hashing` alongside its other `dynamo_tokens` primitives, then import it from there? That lets `dynamo-data-gen` remove the newly reintroduced direct `dynamo-tokens` dependency and keeps all hashing-contract access routed through `dynamo-kv-hashing`.

Comment (RitwijParmar) [reply lib/data-gen/src/mooncake.rs]: Done in `4eda8247fd`.

I re-exported `compute_next_sequence_hash` from `dynamo-kv-hashing`, switched Mooncake to import it from there, and removed the direct `dynamo-tokens` dependency from `dynamo-data-gen` and the lockfile entry. That keeps all data-gen hashing-contract access routed through `dynamo-kv-hashing`.

Revalidated locally:
- `cargo fmt --check --manifest-path lib/data-gen/Cargo.toml`
- `cargo test -p dynamo-data-gen mooncake --lib` -> 23 passed
- `cargo test -p dynamo-kv-hashing consuming_sequence_hashes_match_borrowed_path --test request_hashing`
- `cargo clippy -p dynamo-data-gen --lib --tests -- -D warnings`
- `cargo clippy -p dynamo-kv-hashing --lib --tests -- -D warnings`

File: lib/data-gen/src/mooncake.rs (hunk 3 above)
Comment (PeaBrane) [reply lib/data-gen/src/mooncake.rs]: Current head is still red in Pre Merge because the new `into_blocks_consuming` call in `lib/kv-hashing/src/compute.rs` is not rustfmt-formatted. Could you run repo-wide `cargo fmt --all` plus the relevant clippy checks and push the cleanup so CI gets a clean rerun?

**Comment by RitwijParmar** [comment]
Thanks, good catch. I ran `cargo fmt --all` and pushed the rustfmt-only cleanup in `cb89db2635`.

I also rechecked the relevant local commands before pushing:
- `cargo fmt --all --check`
- `cargo clippy -p dynamo-data-gen --lib --tests -- -D warnings`
- `cargo clippy -p dynamo-kv-hashing --lib --tests -- -D warnings`
- `cargo test -p dynamo-data-gen mooncake --lib`
- `cargo test -p dynamo-kv-hashing consuming_sequence_hashes_match_borrowed_path --test request_hashing`

Will watch the rerun from here.

**Comment by RitwijParmar** [comment]
I checked the new failure from `rust-tests (.)`. It was not in the KV hashing/data-gen path; the failing test was `agents::trace::record::tests::test_emit_request_end_sanitizes_non_finite_metrics`.

The test subscribes to the shared trace bus and was accepting the first `RequestEnd` event it saw. In the full test shard it picked up another request-end record, so the timestamp assertion compared against a live wall-clock timestamp instead of this test’s `1000` value.

I pushed `d85fdaf6b2` to make the test wait for its own `request_id` before asserting. Local checks I could run:
- `cargo fmt --all --check`
- `cargo test -p dynamo-data-gen mooncake --lib`
- `cargo test -p dynamo-kv-hashing consuming_sequence_hashes_match_borrowed_path --test request_hashing`

I also tried the exact `dynamo-llm` test locally, but this machine is missing `protoc`, so `etcd-client` fails during build before the test binary is produced. CI has that dependency, so I’ll watch the rerun there.

**Comment by PeaBrane** [comment]
/ok to test [d85fdaf](https://github.com/ai-dynamo/dynamo/pull/10095/commits/d85fdaf6b2f7486982be4baefa7afb7e4e07acdc)

**Review by PeaBrane — APPROVED** [review]
The requested hashing-contract, dependency-boundary, and consuming-path changes are addressed. Pre-merge Rust tests and clippy are green.
